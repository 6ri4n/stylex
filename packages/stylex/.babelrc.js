function makeHaste() {
  return {
    visitor: {
      ImportDeclaration(path) {
        if (path.get('source').isStringLiteral()) {
          const oldValue = path.get('source').node.value;
          path.get('source').node.value = oldValue.slice(
            oldValue.lastIndexOf('/') + 1
          );
        }
      },
      ExportDefaultDeclaration(path) {
        path.remove();
      },
    },
  };
}

const presets = process.env['HASTE']
  ? []
  : [
      [
        '@babel/preset-env',
        {
          exclude: ['@babel/plugin-transform-typeof-symbol'],
          targets: 'defaults',
        },
      ],
      '@babel/preset-flow',
      '@babel/preset-react',
    ];

const plugins = process.env['HASTE']
  ? [makeHaste, '@babel/plugin-syntax-flow', '@babel/plugin-syntax-jsx']
  : [];

module.exports = {
  assumptions: {
    iterableIsArray: true,
  },
  presets,
  plugins,
};
