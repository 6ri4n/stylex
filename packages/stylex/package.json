{
  "name": "@stylexjs/stylex",
  "version": "0.2.0-beta.12",
  "description": "A styling library for React.",
  "main": "lib/stylex.js",
  "react-native": "lib/native/stylex.js",
  "types": "lib/stylex.d.ts",
  "repository": "https://www.github.com/facebook/stylex",
  "license": "MIT",
  "scripts": {
    "prebuild": "gen-types -i src/ -o lib/",
    "build": "babel src/ --out-dir lib/ --copy-files",
    "build-haste": "HASTE=true babel src/ --out-dir lib/ --copy-files",
    "test": "jest"
  },
  "dependencies": {
    "css-mediaquery": "^0.1.2",
    "invariant": "^2.2.4",
    "styleq": "0.1.3",
    "utility-types": "^3.10.0"
  },
  "devDependencies": {
    "@stylexjs/scripts": "0.2.0-beta.12"
  },
  "jest": {},
  "files": [
    "lib/*"
  ]
}
