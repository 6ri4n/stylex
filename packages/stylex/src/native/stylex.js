/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict
 */

import {
  CSSCustomPropertyValue,
  isCustomPropertyValue,
} from './CSSCustomPropertyValue';
import { CSSLengthUnitValue } from './CSSLengthUnitValue';
import { CSSMediaQuery } from './CSSMediaQuery';
import { errorMsg } from './errorMsg';
import { flattenStyle } from './flattenStyle';
import { parseShadow } from './parseShadow';

const stylePropertyAllowlistSet = new Set<string>([
  'alignContent',
  'alignItems',
  'alignSelf',
  'aspectRatio',
  'backfaceVisibility',
  'backgroundColor',
  'borderBlockColor',
  'borderBlockStyle',
  'borderBlockWidth',
  'borderBlockEndColor',
  'borderBlockEndStyle',
  'borderBlockEndWidth',
  'borderBlockStartColor',
  'borderBlockStartStyle',
  'borderBlockStartWidth',
  'borderBottomColor',
  'borderBottomLeftRadius',
  'borderBottomRightRadius',
  'borderBottomStyle',
  'borderBottomWidth',
  'borderEndEndRadius',
  'borderEndStartRadius',
  'borderColor',
  'borderInlineColor',
  'borderInlineStyle',
  'borderInlineWidth',
  'borderInlineEndColor',
  'borderInlineEndStyle',
  'borderInlineEndWidth',
  'borderInlineStartColor',
  'borderInlineStartStyle',
  'borderInlineStartWidth',
  'borderLeftColor',
  'borderLeftStyle',
  'borderLeftWidth',
  'borderRadius',
  'borderRightColor',
  'borderRightStyle',
  'borderRightWidth',
  'borderStartEndRadius',
  'borderStartStartRadius',
  'borderStyle',
  'borderTopColor',
  'borderTopLeftRadius',
  'borderTopRightRadius',
  'borderTopStyle',
  'borderTopWidth',
  'borderWidth',
  'bottom',
  'color',
  'direction',
  'display',
  'end',
  'flex',
  'flexBasis',
  'flexDirection',
  'flexGrow',
  'flexShrink',
  'flexWrap',
  'fontFamily',
  'fontSize',
  'fontStyle',
  'fontWeight',
  'fontVariant',
  'gap',
  'gapColumn',
  'gapRow',
  'height',
  // 'includeFontPadding', Android Only
  'inset',
  'insetBlock',
  'insetBlockEnd',
  'insetBlockStart',
  'insetInline',
  'insetInlineEnd',
  'insetInlineStart',
  'justifyContent',
  'left',
  'letterSpacing',
  'lineHeight',
  'margin',
  'marginBlock',
  'marginBlockEnd',
  'marginBlockStart',
  'marginBottom',
  'marginInline',
  'marginInlineEnd',
  'marginInlineStart',
  'marginLeft',
  'marginRight',
  'marginStart',
  'marginTop',
  'maxHeight',
  'maxWidth',
  'minHeight',
  'minWidth',
  'objectFit',
  'opacity',
  'overflow',
  'padding',
  'paddingBlock',
  'paddingBlockEnd',
  'paddingBlockStart',
  'paddingBottom',
  'paddingEnd',
  'paddingInline',
  'paddingInlineEnd',
  'paddingInlineStart',
  'paddingLeft',
  'paddingRight',
  'paddingStart',
  'paddingTop',
  'pointerEvents',
  'position',
  'resizeMode',
  'right',
  'shadowColor',
  'shadowOffset',
  'shadowOpacity',
  'shadowRadius',
  'shadowWidth',
  'start',
  'textAlign',
  'textDecorationLine',
  'textDecorationColor', // iOS Only
  'textDecorationStyle', // iOS Only
  'textShadowColor',
  'textShadowOffset',
  'textShadowRadius',
  'textTransform',
  'tintColor',
  'transform',
  'top',
  'userSelect',
  'verticalAlign', // Android Only
  'width',
  'writingDirection', // iOS Only
  'zIndex',
]);

function isReactNativeStyleProp(propName: string): boolean {
  return stylePropertyAllowlistSet.has(propName) || propName.startsWith('--');
}

function isReactNativeStyleValue(propValue: mixed): boolean {
  if (typeof propValue === 'string') {
    // RN doesn't have an inherit keyword
    if (propValue === 'inherit') {
      return false;
    }
    // RN doesn't have an inherit keyword
    if (propValue === 'initial') {
      return false;
    }
    //if (propValue.endsWith('em')) {
    //  return false;
    //}
    //if (propValue.endsWith('rem')) {
    //  return false;
    //}
    // RN on android doesn't like explicitly specified px units
    if (propValue.endsWith('px')) {
      return false;
    }
    // RN doesn't support calc functions
    if (propValue.includes('calc(')) {
      return false;
    }
  }

  return true;
}

function preprocessPropertyValue(propValue: mixed): mixed {
  if (typeof propValue === 'string') {
    if (isCustomPropertyValue(propValue)) {
      return new CSSCustomPropertyValue(propValue);
    }

    const maybeViewportValue = CSSLengthUnitValue.parse(propValue);
    if (maybeViewportValue != null) {
      return maybeViewportValue;
    }
  }

  return propValue;
}

function preprocessCreate<S: { [string]: mixed }>(style: S): S {
  // eslint-disable-next-line flowtype/no-flow-fix-me-comments
  const processedStyle: S = ({}: $FlowFixMe);
  for (const propName in style) {
    const styleValue = style[propName];

    if (
      CSSMediaQuery.isMediaQueryString(propName) &&
      typeof styleValue === 'object' &&
      styleValue != null
    ) {
      // have to spread styleValue into a copied object to appease flow
      const processsedSubStyle = preprocessCreate({ ...styleValue });
      processedStyle[propName] = new CSSMediaQuery(
        propName,
        processsedSubStyle
      );
      continue;
    }

    if (propName === 'backgroundImage') {
      errorMsg('"backgroundImage" is not supported in React Native.');
    }
    // React Native only supports non-standard box-shadow styles
    else if (propName === 'boxShadow' && typeof styleValue === 'string') {
      const parsedShadow = parseShadow(styleValue);
      if (parsedShadow.length > 1) {
        errorMsg(
          'Multiple "boxShadow" values are not supported in React Native.'
        );
      }
      const { inset, offsetX, offsetY, blurRadius, color } = parsedShadow[0];
      // TODO: parse alpha color inputs => alpha + color
      // errorMsg('"boxShadow" opacity is not implemented in React Native.');
      if (inset) {
        errorMsg(
          '"boxShadow" value of "inset" is not supported in React Native.'
        );
      }
      processedStyle.shadowColor = color;
      processedStyle.shadowOffset = { height: offsetY, width: offsetX };
      processedStyle.shadowOpacity = 1;
      processedStyle.shadowRadius = blurRadius;
    } else if (propName === 'position') {
      if (styleValue === 'fixed') {
        processedStyle[propName] = 'absolute';
        errorMsg(
          '"position" value of "fixed" is not supported in React Native. Falling back to "absolute".'
        );
      } else if (styleValue === 'sticky') {
        processedStyle[propName] = 'relative';
        errorMsg(
          '"position" value of "sticky" is not supported in React Native. Falling back to "relative".'
        );
      }
    }
    // React Native only supports non-standard text-shadow styles
    else if (propName === 'textShadow' && typeof styleValue === 'string') {
      const parsedShadow = parseShadow(styleValue);
      if (parsedShadow.length > 1) {
        errorMsg(
          'Multiple "textShadow" values are not supported in React Native.'
        );
      }
      const { offsetX, offsetY, blurRadius, color } = parsedShadow[0];
      processedStyle.textShadowColor = color;
      processedStyle.textShadowOffset = { height: offsetY, width: offsetX };
      processedStyle.textShadowRadius = blurRadius;
    } else {
      processedStyle[propName] = styleValue;
    }
  }

  // Process values that need to be resolved during render
  for (const prop in processedStyle) {
    const processedStyleValue = preprocessPropertyValue(processedStyle[prop]);
    processedStyle[prop] = processedStyleValue;
  }

  return processedStyle;
}

/**
 * The create method shim should do initial transforms like
 * renaming/expanding/validating properties, essentially all the steps
 * which can be done at initialization-time (could potentially be done at
 * compile-time in the future).
 */
export function create<S: { [string]: { ... } }>(styles: S): {
  [string]: { ... },
} {
  const result: { [string]: { ... } } = {};
  for (const styleName in styles) {
    result[styleName] = preprocessCreate(styles[styleName]);
  }
  return result;
}

export const firstThatWorks = <T: string | number>(
  ...values: $ReadOnlyArray<T>
): T => {
  return values[0];
};

export function keyframes(): void {
  errorMsg('keyframes are not supported in React Native.');
}

/**
 * The spread method shim
 */
type SpreadOptions = {|
  customProperties: {},
  inheritedFontSize: ?number,
  fontScale: number | void,
  viewportHeight: number,
  viewportWidth: number,
  writingDirection: 'ltr' | 'rtl',
|};

export function spread(
  style: ?{ [key: string]: mixed },
  {
    customProperties,
    inheritedFontSize,
    fontScale = 1,
    viewportHeight,
    viewportWidth,
    writingDirection,
  }: SpreadOptions
): { [string]: { ... } } {
  /* eslint-disable prefer-const */
  let { lineClamp, ...flatStyle }: { [key: string]: mixed } =
    flattenStyle(style);
  /* eslint-enable prefer-const */

  for (const styleProp in flatStyle) {
    let styleValue = flatStyle[styleProp];

    // resolve media queries
    if (styleValue instanceof CSSMediaQuery) {
      const maybeExistingMediaQuery = flatStyle[styleProp];
      if (maybeExistingMediaQuery instanceof CSSMediaQuery) {
        const s = flattenStyle([
          maybeExistingMediaQuery.matchedStyle,
          styleValue.matchedStyle,
        ]);
        if (s != null) {
          maybeExistingMediaQuery.matchedStyle = s;
        }
        continue;
      }
    }
    // resolve custom property references
    if (styleValue instanceof CSSCustomPropertyValue) {
      const resolvedValue = customProperties[styleValue.name];
      if (resolvedValue == null) {
        errorMsg(`Unrecognized custom property "--${styleValue.name}"`);
        delete flatStyle[styleProp];
        continue;
      }
      styleValue = resolvedValue;
    }
    // resolve viewport units
    if (styleValue instanceof CSSLengthUnitValue) {
      const resolvedValue = styleValue.resolvePixelValue(
        viewportWidth,
        viewportHeight,
        fontScale,
        inheritedFontSize
      );
      styleValue = resolvedValue;
    }

    // Filter out any unexpected style property names so RN doesn't crash but give
    // the developer a warning to let them know that there's a new prop we should either
    // explicitly ignore or process in some way.
    // NOTE: Any kind of prop name transformations should happen before this check.
    if (!isReactNativeStyleProp(styleProp)) {
      errorMsg(`Encountered unsupported style property "${styleProp}"`);
      delete flatStyle[styleProp];
      continue;
    }

    // Similar filter to the prop name one above but instead operates on the property's
    // value. Similarly, any sort of prop value transformations should happen before this
    // filter.
    // We check this at resolve time to ensure the render-time styles are safe.
    if (!isReactNativeStyleValue(styleValue)) {
      errorMsg(
        `Encounted unsupported style value "${String(
          styleValue
        )}" for property "${styleProp}"`
      );
      delete flatStyle[styleProp];
      continue;
    }

    flatStyle[styleProp] = styleValue;
  }

  if (flatStyle != null) {
    flatStyle = CSSMediaQuery.resolveMediaQueries(flatStyle, {
      width: viewportWidth,
      height: viewportHeight,
      direction: writingDirection,
    });
  }

  // Use 'static' position by default for all elements
  if (flatStyle.position == null) {
    flatStyle.position = 'static';
  }

  const nativeProps = { style: flatStyle };

  if (lineClamp != null) {
    // $FlowFixMe
    nativeProps.numberOfLines = lineClamp;
  }

  return nativeProps;
}

export type IStyleX = {
  create: typeof create,
  firstThatWorks: typeof firstThatWorks,
  keyframes: typeof keyframes,
  spread: typeof spread,
};

export const stylex: IStyleX = { create, firstThatWorks, keyframes, spread };
