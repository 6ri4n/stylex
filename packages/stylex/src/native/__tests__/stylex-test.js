/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { stylex } from '../stylex';

const mockOptions = {
  customProperties: {},
  viewportHeight: 600,
  viewportWidth: 320,
};

describe('styles', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error');
    console.error.mockImplementation(() => {});
  });

  afterEach(() => {
    console.error.mockRestore();
  });

  test('background-image', () => {
    const styles = stylex.create({
      root: {
        backgroundImage: 'url(https://placehold.it/300/300',
      },
    });
    stylex.spread(styles.root, mockOptions);
    expect(console.error).toHaveBeenCalled();
  });

  test('box-shadow', () => {
    const styles = stylex.create({
      root: {
        boxShadow: '1px 2px 3px 4px red',
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();

    const styles2 = stylex.create({
      root: {
        boxShadow: '1px 2px 3px 4px red, 2px 3px 4px 5px blue',
      },
    });
    stylex.spread(styles2.root, mockOptions);
    expect(console.error).toHaveBeenCalledTimes(1);

    const styles3 = stylex.create({
      root: {
        boxShadow: 'inset 1px 2px 3px 4px red',
      },
    });
    stylex.spread(styles3.root, mockOptions);
    expect(console.error).toHaveBeenCalledTimes(2);
  });

  test('direction', () => {
    const styles = stylex.create({
      root: {
        direction: 'ltr',
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();

    const styles2 = stylex.create({
      root: {
        direction: 'rtl',
      },
    });
    expect(stylex.spread(styles2.root, mockOptions)).toMatchSnapshot();
  });

  test('font-size', () => {
    const styles = stylex.create({
      root: {
        fontSize: '2.5rem',
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot('default');

    expect(
      stylex.spread(styles.root, { ...mockOptions, fontScale: 2 })
    ).toMatchSnapshot('fontScale:2');
  });

  test('font-variant', () => {
    const styles = stylex.create({
      root: {
        fontVariant: 'common-ligatures small-caps',
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();
  });

  test('font-weight', () => {
    const styles = stylex.create({
      root: {
        fontWeight: 900,
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();

    const styles2 = stylex.create({
      root: { fontWeight: 'bold' },
    });
    expect(stylex.spread(styles2.root, mockOptions)).toMatchSnapshot();
  });

  test('line-clamp', () => {
    const styles = stylex.create({
      root: {
        lineClamp: 3,
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();
  });

  test('object-fit', () => {
    const styles = stylex.create({
      contain: {
        objectFit: 'contain',
      },
      cover: {
        objectFit: 'cover',
      },
      fill: {
        objectFit: 'fill',
      },
      scaleDown: {
        objectFit: 'scale-down',
      },
      none: {
        objectFit: 'none',
      },
    });
    expect(stylex.spread(styles.contain, mockOptions)).toMatchSnapshot(
      'contain'
    );
    expect(stylex.spread(styles.cover, mockOptions)).toMatchSnapshot('contain');
    expect(stylex.spread(styles.fill, mockOptions)).toMatchSnapshot('fill');
    expect(stylex.spread(styles.scaleDown, mockOptions)).toMatchSnapshot(
      'scaleDown'
    );
    expect(stylex.spread(styles.none, mockOptions)).toMatchSnapshot('none');
  });

  test('pointer-events', () => {
    const styles = stylex.create({
      root: {
        pointerEvents: 'none',
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();
  });

  test('position', () => {
    const styles = stylex.create({
      static: {
        position: 'static',
      },
      relative: {
        position: 'relative',
      },
      absolute: {
        position: 'absolute',
      },
      fixed: {
        position: 'fixed',
      },
      sticky: {
        position: 'sticky',
      },
    });
    expect(console.error).toHaveBeenCalledTimes(2);
    expect(stylex.spread(styles.static, mockOptions)).toMatchSnapshot('static');
    expect(stylex.spread(styles.relative, mockOptions)).toMatchSnapshot(
      'relative'
    );
    expect(stylex.spread(styles.absolute, mockOptions)).toMatchSnapshot(
      'absolute'
    );
    expect(stylex.spread(styles.fixed, mockOptions)).toMatchSnapshot('fixed');
    expect(stylex.spread(styles.sticky, mockOptions)).toMatchSnapshot('sticky');
  });

  test('text-shadow', () => {
    const styles = stylex.create({
      root: {
        textShadow: '1px 2px 3px red',
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();

    const styles2 = stylex.create({
      root: {
        textShadow: '1px 2px 3px red, 2px 3px 4px blue',
      },
    });
    expect(stylex.spread(styles2.root, mockOptions)).toMatchSnapshot();
    expect(console.error).toHaveBeenCalledTimes(1);
  });

  test('transform', () => {
    const styles = stylex.create({
      none: {
        transform: 'none',
      },
      matrix: {
        transform: 'matrix(0.1, 1, -0.3, 1, 0, 0)',
      },
      perspective: {
        transform: 'perspective(10px)',
      },
      rotate: {
        transform:
          'rotate(10deg) rotateX(20deg) rotateY(30deg) rotateZ(40deg) rotate3d(0, 0.5, 1, 90deg)',
      },
      scale: {
        transform: 'scale(1, 2) scaleX(1) scaleY(2) scaleZ(3) scale3d(1, 2, 3)',
      },
      skew: {
        transform: 'skew(10deg, 15deg) skewX(20deg) skewY(30deg)',
      },
      translate: {
        transform:
          'translate(10px, 20px) translateX(11px) translateY(12px) translateZ(13px) translate3d(20px, 30px, 40px)',
      },
      mixed: {
        transform: `
          rotateX(1deg) rotateY(2deg) rotateZ(3deg) rotate3d(1deg, 2deg, 3deg)
          scale(1) scaleX(2) scaleY(3) scaleZ(4) scale3d(1,2,3)
          translateX(1px) translateY(1em) translateZ(1rem) translate3d(1px, 1em, 1rem)
        `,
      },
    });
    expect(stylex.spread(styles.none, mockOptions)).toMatchSnapshot('none');
    expect(stylex.spread(styles.matrix, mockOptions)).toMatchSnapshot('matrix');
    expect(stylex.spread(styles.perspective, mockOptions)).toMatchSnapshot(
      'perspective'
    );
    expect(stylex.spread(styles.rotate, mockOptions)).toMatchSnapshot('rotate');
    expect(stylex.spread(styles.scale, mockOptions)).toMatchSnapshot('scale');
    expect(stylex.spread(styles.skew, mockOptions)).toMatchSnapshot('skew');
    expect(stylex.spread(styles.translate, mockOptions)).toMatchSnapshot(
      'translate'
    );
    expect(stylex.spread(styles.rotate, mockOptions)).toMatchSnapshot('rotate');
    expect(stylex.spread(styles.mixed, mockOptions)).toMatchSnapshot('mixed');
  });

  test('user-select', () => {
    const styles = stylex.create({
      root: {
        userSelect: 'none',
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();
  });

  test('vertical-align', () => {
    const styles = stylex.create({
      middle: {
        verticalAlign: 'middle',
      },
      top: {
        verticalAlign: 'top',
      },
    });
    expect(stylex.spread(styles.middle, mockOptions)).toMatchSnapshot('middle');
    expect(stylex.spread(styles.top, mockOptions)).toMatchSnapshot('top');
  });

  test.skip('logical border short-forms', () => {
    const styles = stylex.create({
      root: {},
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();
  });

  test.skip('logical border long-forms', () => {
    const styles = stylex.create({
      root: {},
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();
  });

  test.skip('logical inset short-forms', () => {
    const styles = stylex.create({
      root: {
        insetBlock: 5,
        insetInline: 10,
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();
  });

  test.skip('logical text alignment', () => {
    const styles = stylex.create({
      root: {
        textAlign: 'start',
      },
    });
    expect(stylex.spread(styles.root, mockOptions)).toMatchSnapshot();
  });
});

describe('length units', () => {
  const unitsToTest = ['em', 'px', 'rem', 'vh', 'vmax', 'vmin', 'vw'];
  const value = 10;

  for (const unitToTest of unitsToTest) {
    test(`${value} "${unitToTest}" units are resolved to pixels`, () => {
      const styles = stylex.create({
        underTest: {
          width: `${value}${unitToTest}`,
        },
      });
      expect(stylex.spread(styles.underTest, mockOptions)).toMatchSnapshot();
    });
  }

  test(`${value} "em" units based on inherited font-size`, () => {
    const styles = stylex.create({
      underTest: {
        width: `${value}em`,
      },
    });
    expect(
      stylex.spread(styles.underTest, { ...mockOptions, inheritedFontSize: 12 })
    ).toMatchSnapshot();
  });
});

describe('custom properties', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error');
    console.error.mockImplementation(() => {});
  });

  afterEach(() => {
    console.error.mockRestore();
  });

  test('filters out the property and emits a warning when encountering a variable name which has not been provided', () => {
    const { underTest } = stylex.create({
      underTest: {
        width: 'var(--unprovided)',
      },
    });
    expect(stylex.spread(underTest, mockOptions)).toMatchSnapshot();
    expect(console.error).toHaveBeenCalledWith(
      'stylex: Unrecognized custom property "--unprovided"'
    );
  });

  test("resolves custom properties' values", () => {
    const { underTest } = stylex.create({
      underTest: {
        color: 'var(--slightly-darker-black)',
        width: 'var(--the-best-width)',
      },
    });
    expect(
      stylex.spread(underTest, {
        ...mockOptions,
        customProperties: {
          slightlyDarkerBlack: '#333',
          theBestWidth: 42,
        },
      })
    ).toMatchSnapshot();
  });
});

expect.extend({
  toMatchWindowDimensions(query, windowSize) {
    const { height, width } = windowSize;
    const EXPECTED_MATCHED_VALUE = 500;
    const { underTest } = stylex.create({
      underTest: {
        [`@media ${query}`]: {
          width: EXPECTED_MATCHED_VALUE,
        },
      },
    });
    const props = stylex.spread(underTest, {
      viewportHeight: height,
      viewportWidth: width,
    });
    const actualValue = props.style.width;
    if (actualValue === EXPECTED_MATCHED_VALUE) {
      return {
        pass: true,
        message: () =>
          `expected media query ${query} not to match in a window of width ${width} and height ${height}`,
      };
    } else {
      return {
        pass: false,
        message: () =>
          `expected media query ${query} to match in a window of width ${width} and height ${height}`,
      };
    }
  },
});

describe('media queries', () => {
  test('matches a "min-width" query', () => {
    expect('(min-width: 400px)').toMatchWindowDimensions({
      width: 450,
      height: 0,
    });
    expect('(min-width: 400px)').not.toMatchWindowDimensions({
      width: 350,
      height: 0,
    });
  });

  test('matches a "max-width" query', () => {
    expect('(max-width: 400px)').toMatchWindowDimensions({
      width: 350,
      height: 0,
    });
    expect('(max-width: 400px)').not.toMatchWindowDimensions({
      width: 450,
      height: 0,
    });
  });

  test('matches a "min-height" query', () => {
    expect('(min-height: 400px)').toMatchWindowDimensions({
      width: 0,
      height: 450,
    });
    expect('(min-height: 400px)').not.toMatchWindowDimensions({
      width: 0,
      height: 350,
    });
  });

  test('matches a "max-height" query', () => {
    expect('(max-height: 400px)').toMatchWindowDimensions({
      width: 0,
      height: 350,
    });
    expect('(max-height: 400px)').not.toMatchWindowDimensions({
      width: 0,
      height: 450,
    });
  });
});

describe('unsupported style values', () => {
  beforeEach(() => {
    jest.spyOn(console, 'error');
    console.error.mockImplementation(() => {});
  });

  afterEach(() => {
    console.error.mockRestore();
  });

  test('calc', () => {
    const { underTest } = stylex.create({
      underTest: {
        width: 'calc(2 * 1rem)',
      },
    });
    expect(stylex.spread(underTest, mockOptions)).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();
  });

  test('inherit', () => {
    const { underTest } = stylex.create({
      underTest: {
        fontSize: 'inherit',
      },
    });
    expect(stylex.spread(underTest, mockOptions)).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();
  });

  test('initial', () => {
    const { underTest } = stylex.create({
      underTest: {
        fontSize: 'initial',
      },
    });
    expect(stylex.spread(underTest, mockOptions)).toMatchSnapshot();
    expect(console.error).toHaveBeenCalled();
  });
});
