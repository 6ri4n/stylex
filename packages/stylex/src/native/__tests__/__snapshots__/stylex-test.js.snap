// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`custom properties filters out the property and emits a warning when encountering a variable name which has not been provided 1`] = `{}`;

exports[`custom properties resolves custom properties' values 1`] = `
{
  "style": {
    "color": "#333",
    "width": 42,
  },
}
`;

exports[`length units 10 "em" units are resolved to pixels 1`] = `
{
  "style": {
    "width": 160,
  },
}
`;

exports[`length units 10 "em" units based on inherited font-size 1`] = `
{
  "style": {
    "width": 120,
  },
}
`;

exports[`length units 10 "px" units are resolved to pixels 1`] = `
{
  "style": {
    "width": 10,
  },
}
`;

exports[`length units 10 "rem" units are resolved to pixels 1`] = `
{
  "style": {
    "width": 160,
  },
}
`;

exports[`length units 10 "vh" units are resolved to pixels 1`] = `
{
  "style": {
    "width": 60,
  },
}
`;

exports[`length units 10 "vmax" units are resolved to pixels 1`] = `
{
  "style": {
    "width": 60,
  },
}
`;

exports[`length units 10 "vmin" units are resolved to pixels 1`] = `
{
  "style": {
    "width": 32,
  },
}
`;

exports[`length units 10 "vw" units are resolved to pixels 1`] = `
{
  "style": {
    "width": 32,
  },
}
`;

exports[`styles box-shadow 1`] = `
{
  "style": {
    "shadowColor": "red",
    "shadowOffset": {
      "height": 2,
      "width": 1,
    },
    "shadowOpacity": 1,
    "shadowRadius": 3,
  },
}
`;

exports[`styles direction 1`] = `
{
  "style": {
    "direction": "ltr",
  },
}
`;

exports[`styles direction 2`] = `
{
  "style": {
    "direction": "rtl",
  },
}
`;

exports[`styles font-size: default 1`] = `
{
  "style": {
    "fontSize": 40,
  },
}
`;

exports[`styles font-size: fontScale:2 1`] = `
{
  "style": {
    "fontSize": 80,
  },
}
`;

exports[`styles font-variant 1`] = `
{
  "style": {
    "fontVariant": "common-ligatures small-caps",
  },
}
`;

exports[`styles font-weight 1`] = `
{
  "style": {
    "fontWeight": "900",
  },
}
`;

exports[`styles font-weight 2`] = `
{
  "style": {
    "fontWeight": "bold",
  },
}
`;

exports[`styles line-clamp 1`] = `
{
  "numberOfLines": 3,
}
`;

exports[`styles object-fit: contain 1`] = `
{
  "style": {
    "objectFit": "contain",
  },
}
`;

exports[`styles object-fit: contain 2`] = `
{
  "style": {
    "objectFit": "cover",
  },
}
`;

exports[`styles object-fit: fill 1`] = `
{
  "style": {
    "objectFit": "fill",
  },
}
`;

exports[`styles object-fit: none 1`] = `
{
  "style": {
    "objectFit": "none",
  },
}
`;

exports[`styles object-fit: scaleDown 1`] = `
{
  "style": {
    "objectFit": "scale-down",
  },
}
`;

exports[`styles pointer-events 1`] = `
{
  "style": {
    "pointerEvents": "none",
  },
}
`;

exports[`styles position: absolute 1`] = `
{
  "style": {
    "position": "absolute",
  },
}
`;

exports[`styles position: fixed 1`] = `
{
  "style": {
    "position": "absolute",
  },
}
`;

exports[`styles position: relative 1`] = `
{
  "style": {
    "position": "relative",
  },
}
`;

exports[`styles position: static 1`] = `
{
  "style": {
    "position": "static",
  },
}
`;

exports[`styles position: sticky 1`] = `
{
  "style": {
    "position": "relative",
  },
}
`;

exports[`styles text-shadow 1`] = `
{
  "style": {
    "textShadowColor": "red",
    "textShadowOffset": {
      "height": 2,
      "width": 1,
    },
    "textShadowRadius": 3,
  },
}
`;

exports[`styles text-shadow 2`] = `
{
  "style": {
    "textShadowColor": "red",
    "textShadowOffset": {
      "height": 2,
      "width": 1,
    },
    "textShadowRadius": 3,
  },
}
`;

exports[`styles transform: matrix 1`] = `
{
  "style": {
    "transform": "matrix(0.1, 1, -0.3, 1, 0, 0)",
  },
}
`;

exports[`styles transform: mixed 1`] = `
{
  "style": {
    "transform": "
          rotateX(1deg) rotateY(2deg) rotateZ(3deg) rotate3d(1deg, 2deg, 3deg)
          scale(1) scaleX(2) scaleY(3) scaleZ(4) scale3d(1,2,3)
          translateX(1px) translateY(1em) translateZ(1rem) translate3d(1px, 1em, 1rem)
        ",
  },
}
`;

exports[`styles transform: none 1`] = `
{
  "style": {
    "transform": "none",
  },
}
`;

exports[`styles transform: perspective 1`] = `
{
  "style": {
    "transform": "perspective(10px)",
  },
}
`;

exports[`styles transform: rotate 1`] = `
{
  "style": {
    "transform": "rotate(10deg) rotateX(20deg) rotateY(30deg) rotateZ(40deg) rotate3d(0, 0.5, 1, 90deg)",
  },
}
`;

exports[`styles transform: rotate 2`] = `
{
  "style": {
    "transform": "rotate(10deg) rotateX(20deg) rotateY(30deg) rotateZ(40deg) rotate3d(0, 0.5, 1, 90deg)",
  },
}
`;

exports[`styles transform: scale 1`] = `
{
  "style": {
    "transform": "scale(1, 2) scaleX(1) scaleY(2) scaleZ(3) scale3d(1, 2, 3)",
  },
}
`;

exports[`styles transform: skew 1`] = `
{
  "style": {
    "transform": "skew(10deg, 15deg) skewX(20deg) skewY(30deg)",
  },
}
`;

exports[`styles transform: translate 1`] = `
{
  "style": {
    "transform": "translate(10px, 20px) translateX(11px) translateY(12px) translateZ(13px) translate3d(20px, 30px, 40px)",
  },
}
`;

exports[`styles user-select 1`] = `
{
  "style": {
    "userSelect": "none",
  },
}
`;

exports[`styles vertical-align: middle 1`] = `
{
  "style": {
    "verticalAlign": "middle",
  },
}
`;

exports[`styles vertical-align: top 1`] = `
{
  "style": {
    "verticalAlign": "top",
  },
}
`;

exports[`unsupported style values calc 1`] = `{}`;

exports[`unsupported style values inherit 1`] = `{}`;

exports[`unsupported style values initial 1`] = `{}`;
